In [1]:
x = [1,2,3,4]
x

[1, 2, 3, 4]

In [2]:
a = [1,2,3,4]
b = [2,3,4,5]
a+b

[1, 2, 3, 4, 2, 3, 4, 5]

Imports:

In [3]:
import tensorflow as tf
print(tf.__version__)

2.21.0


In [4]:
import numpy as np
# import matplotlib.pyplot as plt
# from mpl_toolkits.mplot3d import Axes3D

Generate inputs:

In [5]:
observations = 1000

#two variable linear model
xs = np.random.uniform(low=-10, high=10,size=(observations,1))
zs = np.random.uniform(low=-10, high=10,size=(observations,1))
#print(xs,zs)

inputs = np.column_stack((xs,zs))
print(inputs.shape)

(1000, 2)


Create targets:

In [6]:
#targets = f(x,z) = 2x - 3z +5 + noise
noise = np.random.uniform(-1,1,size=(observations,1))
targets = 2*xs - 3*zs + 5 + noise

print(targets.shape)

(1000, 1)


Plot results:

Initialize variables:

In [7]:
init_range = 0.1
weights = np.random.uniform(-init_range, init_range,size=(2,1))
biases = np.random.uniform(-init_range, init_range,size=1)

print(weights, biases)

print(weights.shape, biases.shape)


[[ 0.07716437]
 [-0.0631375 ]] [-0.01707356]
(2, 1) (1,)


Learning rate:

In [8]:
learning_rate = 0.001

Training:

    Plan:
        - calculate outputs
        - compare outputs to targets through loss
        - print loss
        - Adjust weights and biases
        - Repeat ~

In [9]:
for i in range(10000):
    outputs = np.dot(inputs,weights) + biases
    deltas = outputs - targets

    #L2-norm loss formula -> Ei(yi-ti)squared 2
    loss = np.sum(deltas**2)/2/observations

    print(loss)

    deltas_scaled=deltas / observations

    #print(delta_scaled.shape)

    weights = weights - learning_rate * np.dot(inputs.T,deltas_scaled)
    biases = biases - learning_rate * np.sum(deltas_scaled)

221.26385071076155
207.5188307575868
194.67654075338115
182.67756312617462
171.46639322777202
160.99118152752698
151.20349279980746
142.058081184475
133.51268007363333
125.52780584696576
118.06657454247674
111.0945306096988
104.57948694869194
98.49137549071636
92.80210762554479
87.4854438262262
82.51687186493355
77.8734930535231
73.5339159797874
69.47815724527614
65.68754874314631
62.14465104494229
58.83317249363528
55.737893626806866
52.844596578663065
50.13999913273138
47.611693118730926
45.24808686731633
43.03835145527297
40.97237049137282
39.040693209569376
37.23449065159216
35.54551473536987
33.96606001913055
32.488927983563485
31.10739366613477
29.815174492585463
28.60640116085619
27.475590442223105
26.417619773342544
25.427703521226366
24.50137081094504
23.63444481311905
22.82302339504271
22.063461045621498
21.352351990222704
20.686514417068164
20.06297574196174
19.47895884296828
18.9318692011663
18.419282887805608
17.93893534213208
17.48871088781446
17.0666329393376
16.67085485

In [10]:
print(weights, biases)

[[ 2.00348959]
 [-2.9978845 ]] [5.00537985]


Result 1: [[ 2.00076795]
 [-3.00090749]] [5.01994943]

 After about 400 iterations the loss seems to have settled at 0.16 at the learning rate of 0.01.

 This results seems to be same even with increased iterations.
 Lets slow down the learning rate to 0.001
 Result 2: [[ 2.01088028]
 [-3.01838722]] [3.1928732]
  After about 1000 iterations the loss seems to have processed only until 1.79 at the learning rate of 0.001.
The slow learning rate needed more time(iterations). lets try with increasing the iterations to 2000.

Result 3: [[ 1.99975776]
 [-3.00520799]] [4.33073023]

 Close enough to the first, but the loss could reach only until 0.39.
 We have two options- increase the iterations to arrive at the loss close to 0.1 or increase learning rate(this may be ideal as we dont see much benefit in slow learning and high iterations)

But first lets try increase the iteration to 10000 and see.
 Note: This might take very long time compared to initial ones taking ~10 secs. Now the question is, is it worth the load on the machine with limited resources OR shall we run this on a GPU ?!?!  Welcome to the Work of GPUs.

 In the real world scenerios, we train models thousands of magnitude of this with million of data and that needs multiple runs(trainings) to arrive at the perfect Model with several parameters(here we just have 2). In which case normal Machine will be very slow and are not meant for the job. Hence the GPUs solve the problems.
 
Psst.... while I writing this the training process is still running at 300 seconds!! Also notice that the value is kind of settled at 0.17 with very less progress tending towards 0.1. The question is, is it worth it?

The process is more about otimizing the available resources to the generating that perfect Model.

Finally the loss ended at exactly 0.17218435462997075.

Result4: [[ 2.00348959]
 [-2.9978845 ]] [5.00537985]

 We could arrive at ideal solution being in Result 1, but we couldn't have concluded it without the multiple training runs .. :-)


